content/post/2023/14-solving-a-sudoku-with-symbiyosys-and-formal-verification.md

```diff
@@ -0,0 +1,305 @@
+```verilog
+generate
+	for(i = 0; i < 9; i = i + 1) begin: row_check
+		for(j = 0; j < 9; j = j + 1) begin: column_j
+			for(k = j + 1; k < 9; k = k + 1) begin: column_k
+				always @(posedge clk) begin
+					assume(sudoku_grid[i][j] !== sudoku_grid[i][k]);
+				end
+			end
+		end
+	end
+endgenerate
+```
+
+3. **All digits in a column are all different**
+```verilog
+generate
+	for(i = 0; i < 9; i = i + 1) begin: column_check
+		for(j = 0; j < 9; j = j + 1) begin: row_j
+			for(k = j + 1; k < 9; k = k + 1) begin: row_k
+				always @(posedge clk) begin
+					assume(sudoku_grid[j][i] !== sudoku_grid[k][i]);
+				end
+			end
+		end
+	end
+endgenerate
+```
+
+4. **All digits in a box are all different**
+```verilog
+generate
+	for(box_row = 0; box_row < 3; box_row = box_row + 1) begin: box_row_check
+		for(box_col = 0; box_col < 3; box_col = box_col + 1) begin: box_col_check
+			for(i = 0; i < 3; i = i + 1) begin: row_i_check
+				for(j = 0; j < 3; j = j + 1) begin: col_j_check
+					for(m = 0; m < 3; m = m + 1) begin: row_m_check
+						for(n = 0; n < 3; n = n + 1) begin: col_n_check
+						// Make sure we're not comparing the same cell to itself
+							if(i !== m || j !== n) begin 
+								always @(posedge clk) begin
+assume(sudoku_grid[(3*box_row)+i][(3*box_col)+j] !== sudoku_grid[(3*box_row)+m][(3*box_col)+n]);
+								end
+							end
+						end
+					end
+				end
+			end
+		end
+	end
+endgenerate
+```
+
+5. **The initial grid cannot be changed**. 
+``` verilog
+// assume the intial grid
+always @(posedge clk) begin : initialization
+//line 1
+assume(sudoku_grid[0][0]==5);
+assume(sudoku_grid[0][2]==7);
+assume(sudoku_grid[0][3]==2);
+assume(sudoku_grid[0][7]==9);
+//line 2
+assume(sudoku_grid[1][2]==6);
+assume(sudoku_grid[1][5]==3);
+assume(sudoku_grid[1][6]==7);
+assume(sudoku_grid[1][8]==1);
+//line 3
+assume(sudoku_grid[2][0]==4);
+assume(sudoku_grid[2][7]==6);
+//line 4
+assume(sudoku_grid[3][0]==1);
+assume(sudoku_grid[3][3]==4);
+assume(sudoku_grid[3][4]==9);
+assume(sudoku_grid[3][8]==7);
+//line 5
+assume(sudoku_grid[4][3]==5);
+assume(sudoku_grid[4][5]==8);
+//line 6
+assume(sudoku_grid[5][0]==8);
+assume(sudoku_grid[5][4]==2);
+assume(sudoku_grid[5][5]==7);
+assume(sudoku_grid[5][8]==5);
+//line 7
+assume(sudoku_grid[6][1]==7);
+assume(sudoku_grid[6][8]==9);
+//line 8
+assume(sudoku_grid[7][0]==2);
+assume(sudoku_grid[7][2]==9);
+assume(sudoku_grid[7][4]==8);
+assume(sudoku_grid[7][6]==6);
+//line 9
+assume(sudoku_grid[8][1]==4);
+assume(sudoku_grid[8][5]==9);
+assume(sudoku_grid[8][6]==3);
+assume(sudoku_grid[8][8]==8);
+end
+```
+
+### Hijacking the use of the SymbiYosys
+
+With the game rules handed over to our solver, we want it to return the solution now. Here lies our ingenious twist in using the tool!
+
+As highlighted before, in a correctly solved Sudoku, every row, column, or box's digit sum is 45. If we were to declare: "assert(sum_line=6'd45);" the solver would merely confirm with a "yes"... and leave it at that.
```

Comment:
Since you are actually looking for an example and not a counter-example it might be a good idea to use SBY's `mode cover` with `cover(sum==6'd45);` instead of `assert(sum!=6'd45)`. It's more or less equivalent but "example to what I want" is more to the point than "counter-example to what I don't want". (Also I think `cover(1)`, i.e. "show me any example for my assumptions" should work as well since your assumptions are already enough to make sure the solution has that sum.)

Reply:
Thanks again, I am new to SBY. I have implemented the cover solution.